File: LeetCode/Explore/TopInterview/Blind75/Tree/98_validate_binary_search_tree.cpp
// Recursive Traversal with Valid Range

// Complexity Analysis
// Time complexity : \mathcal{O}(N)O(N) since we visit each node exactly once.
// Space complexity : \mathcal{O}(N)O(N) since we keep up to the entire tree.

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
  bool isValidBST(TreeNode* root, long Min = LONG_MIN, long Max = LONG_MAX) {
    if (!root) { // base case
      return true;
    }
    if (root->val <= Min || root->val >= Max) {
      return false;
    }
    return isValidBST(root->left, Min, root->val) && isValidBST(root->right, root->val, Max);
  }
};

File: LeetCode/Explore/TopInterview/Blind75/Tree/98_ validate_binary_search_tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode], MIN=-math.inf, MAX=math.inf) -> bool:
        if root == None:
          return True
        
        if root.val <= MIN or root.val >= MAX:
          return False
        
        return self.isValidBST(root.left, MIN, root.val) and self.isValidBST(root.right, root.val, MAX)
